### DISCUSSION TITLE
v4

### DISCUSSION THREAD

**Discussion by RussKie** [gitextensions/gitextensions#10310] [General]
### Release Notes Highlights

* The app now requires .NET 6. The .NET runtime, if missing, will be installed when the application is executed.
* Improved repository load time considerably (#9243, #9735, #9864, etc) as well as other performance tweaks.
* Improved "left panel" with the added multi-selection functionality (#9723) and the ability to view git stashes (#10258).
* Improved filtering functionality including the ability to "exclude" git refs (#10240).
* Use native WSL Git executables for `\\wsl$` shares to speedup handling (#9702)
* Use the standard `FormBrowse` for file history and blame (#9445)
* BASE diff: Icons for unique changes (#9720)

### The following features and options have been removed:
* PuTTY binaries are no longer bundled with Git Extentions, please install PuTTY separately and change the path in settings.
* Theming is not supported - v3.x relied on EasyHook library to facilitate hooks into Windows API necessary for theming. However, .NET 6 does not support remoting and multiple AppDomain required for those hooks. To date no viable alternative has been found. https://github.com/gitextensions/gitextensions/pull/9191
* Removed TFS integration (https://github.com/gitextensions/gitextensions/pull/8522/commits/8190d0c0570af95ffbf9e168072d1b169eb99618)
* Removed JIRA integration (see: https://github.com/gitextensions/gitextensions/issues/9659)

![Downloads](https://img.shields.io/github/downloads/gitextensions/gitextensions/v4/total.svg)

<hr /><em>This discussion was created from the release <a href='https://github.com/gitextensions/gitextensions/releases/tag/v4'>v4</a>.</em>

**Comment by Tampa** [comment]
No dark mode no download. While themes are nice, at least a dark mode with forms is possible. Yes it needs to directly change backgrounds and colors requiring it's own project, but Blinded by the light should remain a song lyric.

**Reply by pmiossec** [reply]
> No dark mode no download.

No download, no right to give it's point of view, right!?! In fact, we don't care if annoying people doesn't download our software. We prefer to talk with people understanding the lot of time we spend on it for free... 

>While themes are nice, at least a dark mode with forms is possible. Yes it needs to directly change backgrounds and colors requiring it's own project, but Blinded by the light should remain a song lyric.

We always appreciate a comment where the only goal is to tell us to do a huge work (on our free time). 

But as you got the idea first and you seem to know what to do, please do a PR (and stay around to maintain it)...

**Reply by Tampa** [reply]
Someone feels triggered. If you don't want to work on a project then don't, but when you release a piece of software into the wild expect criticism, especially on regressions, even more so regarding dark mode(one of the things that gets requested for any gui app the moment it's available for download). It's not a huge deal to duplicate the project file and apply a patch to change the colors, but of course, as I said, that's not solving the underlying issue, just provides a hack to have some form of dark mode. Plus it's a b**** to maintain and setup every single background and color, not disputing that.

That dotnet drags its feet on pretty much everything is nothing new either, it's caused issues for everything that depends on it, so the blame is squarely on them. Deprecating features without replacements or even workarounds is not good practice, but that should not discourage anyone from seeking them. The discussion on the topic seems to center around trying to make it work with existing codebase, when it might just be time to ditch the idea of trying to make pigs fly and seek a replacement strategy for a more versatile and less volatile system from the frontend. Course that's a lot of work, but it's unlikely it'll ever not be, not with the course dotnet is taking.

Regressions are never fun to deal with, especially if they involve a lot of work to resolve, but in this case I would have elected to perhaps classify this release as a beta until such time a plan is formulated what to do about themes or at least introducing one of most requested features of any gui app, dark mode, back into it in some form, even just as a hack.

**Reply by pmiossec** [reply]
> Someone feels triggered. If you don't want to work on a project then don't, but when you release a piece of software into the wild expect criticism, especially on regressions, even more so regarding dark mode(one of the things that gets requested for any gui app the moment it's available for download).

Feel reassured that I have not waited for your words to have a real life but don't put the blame on someone else on your bad behavior. 
There 2 ways to give feedback and do criticism. And the one you did was not constructive nor kind (to stay polite). 

If you think it's easy to do and maintain, do a PR. We will love to give you feedback to be able to achieve it.

It's easy to talk without a line written on such a feature where the devil is in the details and where a lot of edge cases will surely will have to be handled.

It's easy to come just after the release with your point of view but where were you during nearly the 2 years of development of this version? When the alpha was released? And the beta? Your PR would have the time to be merged since a long time... 
Did you read at least a line of the discussions around the decision to release?

Lots of talk... Now that i'm sure that it will be easy for you to do the change, I'm happy to wait for the PR because I really like to have a dark version also.

**Reply by Tampa** [reply]
I been following the issue tickets regarding the frontend issues and have had the misfortune to deal with dotnet6 myself, so I am very much aware of the annoyance that is to bring code up to. It's funny you assume I am negative towards this release. I merely stated a sentiment that a lack of dark mode makes me not want to download this version specifically. I am, as I said in my longer reply, full aware of the complexity involved in even hacking it back in.

The part that irks me is the idea that the solution to the themeing removal would be something coming along down the line that would allow an easy path to re-introduce it, when it's pretty clear from the way MS handles dotnet that they are not interested to provide such paths for anything that has been removed. This is why I mentioned just hacking together a dark mode instead of attempting full theme support.

MS killed mono, removed various things from dotnet without any replacements or even workarounds. The developers working on dotnet have no interest in projects asking for deprecated features to be implemented in a way. The aim seems squarely to support corporate interests and, frankly I don't even know what the whole idea is, but it's a bit of a directionless mess right now. Perhaps this changes down the line, but I know better not to depend on this and just make it work another way. :)

**Reply by RussKie** [reply]
@Tampa thank you for your feedback. Whilst you're absolutely entitled to your opinion, it's largely unwanted because you express it in inflammatory manner, and it contains a number of generalisations, assumptions and accusations which have no basis. Irrespective of what your think of Microsoft, .NET or whatever other technology may be used in this project, coming here and ranting about it is inappropriate. For any issues with .NET direct your feedback to the .NET team. Git Extensions comes "AS IS" with all its features, deficiencies and bugs. The members of the team and external contributors have been selflessly contributing to the project with the aim to make it better. If nothing else, you should be thanking them for this.

**Reply by lesair** [reply]
If no one has bandwidth to restore back the dark theme working on v4 I volunteer to do it.

**Reply by pmiossec** [reply]
@lesair That would be great!
FYI, there is an issue on the subject if you want to continue the discussion there: #9191

**Reply by glen-nicol** [reply]
Just for documentation purposes, without theming support, is it possible to change the branch colors in v4? The settings box still has UI related to themes which is confusing. I really hate that red and green are swapped from the git CLI default. That is more annoying than missing dark mode to me.

This didn't seem as applicable to the #9191 issue where they are discussing how to bring theming back.

**Reply by pmiossec** [reply]
> Just for documentation purposes, without theming support, is it possible to change the branch colors in v4?

That's still supported. In user themes folder

![image](https://user-images.githubusercontent.com/460196/200399980-c7a6d93c-7b29-43ee-ace2-0ac20bd3db5e.png)


Create a `.css` file with the content:

```
@import url("invariant.css");
.Branch                  { color: #e66363; }
.RemoteBranch            { color: #22a639; }
```

**Reply by glen-nicol** [reply]
\\(ᵔᵕᵔ)/ Thank you thank you!

**Comment by bcdiaconu** [comment]
@RussKie: will the winget package be upgraded?

**Reply by RussKie** [reply]
I'm not aware that the team maintains the winget or chocolatey packages, sorry.

**Reply by pmiossec** [reply]
@bcdiaconu I think you could do a PR in this repository ( https://github.com/microsoft/winget-pkgs/tree/master/manifests/g/GitExtensionsTeam/GitExtensions ) following these steps: https://github.com/microsoft/winget-pkgs#submitting-a-package
by doing something similar to the previous version https://github.com/microsoft/winget-pkgs/pull/29099

Or open an issue in the same repository to ask for a "package request". Maybe your request will be eared by someone...

**Comment by davidsmoot** [comment]
I was prompted this morning in the app to update.  The link in the prompt to "Show Changelog" does not reference 4.0.0 at all.  Which led me to be a little suspicious until I found this page.  Might want to consider updating the readme at /gitextensions/gitextensions/blob/master/GitUI/Resources/ChangeLog.md to reflect 4.0 updates.

**Reply by RussKie** [reply]
Thank you, completely forgotten about this one. #10393

**Comment by ATS-CE** [comment]
Are there any migrations needed to move from 3.5.4 to 4.0.0? Or simply uninstall 3.5.4 and install 4.0.0?

**Reply by vbjay** [reply]
If using putty you need to install it and point paths to installed versions.

**Comment by LightCC** [comment]
I uninstalled 3.5.4, then installed 4.0.0. I selected the "Install just for you" option, but on the next screen it was still showing me an install folder of Program Files rather than my home directory or appdata, etc. I continued installing and that's where it installed it.

In addition, the installer doesn't check on the dot net (which was the impression I got from the release notes) - that doesn't happen until you try to run the app itself. Not sure if that step requires an admin, but if so, it's important for folks that need IT to install when admin is needed (I have admin, but others in our company do not). Not sure it needs to be moved, but the release notes/docs should be clear.

**Reply by RussKie** [reply]
I don't think per-user install ever worked.

**Comment by IlyaMokin** [comment]
link to .net 6.0.9 is wrong.
It shows a modal like "please install .net runtime 6.0.9" but by clicking the link it opens download page for most recent version of .net which is **6.0.11**. Should be fine probably, but it do not allow to run gitextensions under 6.0.11 and showing the same modal again... So the correct link is this one
https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/runtime-desktop-6.0.9-windows-x64-installer?cid=getdotnetcore

**Reply by vbjay** [reply]
It is ok to have a patched version. Later versions are ok.  You must have at least 6.0.9 desktop runtime installed.

**Reply by IlyaMokin** [reply]
@vbjay, technically I believe you are right. But it doesn't work with 6.0.11. Just showing a modal that we need to install 6.0.9

**Reply by vbjay** [reply]
Just as an info gathering thing. 

Can you run 

```cmd

dotnet --list-runtimes

dotnet --list-sdks
```

**Reply by IlyaMokin** [reply]
I installed 6.0.9 after it failed with 6.0.11. With 6.0.9 everything is ok.

My current output is this one, but probably it says nothing as both are installed for me now
`
C:\Users\iliam>dotnet --list-runtimes
Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

C:\Users\iliam>dotnet --list-sdks
5.0.101 [C:\Program Files\dotnet\sdk]
6.0.100 [C:\Program Files\dotnet\sdk]
6.0.302 [C:\Program Files\dotnet\sdk]

**Reply by PaoloOranges** [reply]
I confirm what @IlyaMokin said. I installed the link provided when opening GitExtensions and it downloads dotnet runtime 6.0.11. GitExtensions doesn't work with that one.
Here is the output of ```dotnet --list-runtimes``` before installing 6.0.9 but after 6.0.11.

```
PS C:\Program Files (x86)\GitExtensions> dotnet --list-runtimes
Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 5.0.0-preview.4.20257.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 5.0.0-preview.4.20251.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 5.0.0-preview.4.20251.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Reply by vbjay** [reply]
Pay attention. It needs the **desktop runtime**

**Reply by PaoloOranges** [reply]
> Pay attention. It needs the **desktop runtime**

True that. This is the correct link: [https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/runtime-desktop-6.0.11-windows-x64-installer](url)

**Reply by pmiossec** [reply]
@RussKie Maybe it could be a good idea to edit the first post and add a first line like:

>To be able to launch GitExtensions, you will have to install **.NET Desktop Runtime** >= 6.0.9 that you could find following [this link](https://dotnet.microsoft.com/en-us/download/dotnet/6.0#runtime-desktop-6.0.11)

What do you think?

**Reply by orSolocate** [reply]
I installed these 2 which enabled launching the new GitExtensions.

https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/runtime-aspnetcore-6.0.11-windows-x64-installer
https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-6.0.403-windows-x64-installer

Without installing it manually the nothing will be launched when opening the `.exe`. Windows 10 Pro, x64 ARCH, 64 bit OS.

**Reply by RussKie** [reply]
This was fixed in https://github.com/gitextensions/gitextensions/pull/10355

**Reply by ChaerilM** [reply]
```
PS C:\WINDOWS\system32> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.101
 Commit:    bc5f8df0f5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.101\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  3.0.101 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info
```

got desktop runtime, but did not work, does it wants the SDK instead ?

**Reply by RussKie** [reply]
Only .NET Desktop runtime is required. What errors are you getting?

**Reply by ChaerilM** [reply]
same error, need desktop runtime 6.0.9, but after installing it, it does not work. its working after i installed sdk
```
.NET SDK (reflecting any global.json):
 Version:   6.0.403
 Commit:    2bc18bf292

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.403\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  3.0.101 [C:\Program Files\dotnet\sdk]
  6.0.403 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
``

**Reply by vbjay** [reply]
Sounds more like an x64/x86 difference. Sdk installs both if x64.

**Comment by chenchaohan** [comment]
没有汉化版本吗?

**Comment by mstv** [comment]
> 没有汉化版本吗?

LMGTFY: "No Chinese version?"

This is a volunteer project. Feel free to contribute: https://github.com/gitextensions/gitextensions/wiki/Translations#how-to-update-non-english-translations.

**Reply by RussKie** [reply]
Chinese translations weren't available when v4 release was assembled. I think those will shipped in v4.0.1.
